File: java/src/main/java/S051.java
import com.google.common.collect.Lists;
import com.google.common.collect.Sets;

import java.util.List;
import java.util.Set;

// TODO: way too bespoke
public class S051 {

    private static final int LIMIT = 1000 * 1000;

    public static int solve() {
        Set<Integer> primes = Sets.newLinkedHashSet(Maths.generatePrimesBelow(LIMIT));
        for (int prime : primes) {
            int[][] digitGroups = getDigitGroups(prime);
            for (int[] group : digitGroups) {
                maskLoop: for (int mask = 1; mask < (1 << group.length); mask++) {
                    List<Integer> resultingPrimes = Lists.newArrayList();
                    for (int replaceWith = 0; replaceWith < 10; replaceWith++) {
                        int replaced = replace(prime, group, mask, replaceWith);
                        if (primes.contains(replaced)) {
                            resultingPrimes.add(replaced);
                        }
                        if (resultingPrimes.size() < replaceWith - 1) {
                            continue maskLoop;
                        }
                    }
                    if (resultingPrimes.size() == 8) {
                        return resultingPrimes.get(0);
                    }
                }
            }
        }
        return -1;
    }

    private static int replace(int original, int[] locations, int mask, int replacementDigit) {
        int[] digits = Maths.getDigits(original);
        for (int locIdx = 0; locIdx < locations.length; locIdx++) {
            int replacementIdx = locations[locIdx];
            // Do not replace first digit with 0
            if (replacementIdx == 0 && replacementDigit == 0) {
                return -1;
            }
            if ((mask >>> locIdx) % 2 == 1) {
                digits[replacementIdx] = replacementDigit;
            }
        }
        return (int) Maths.digitToLong(digits);
    }

    private static int[][] getDigitGroups(int n) {
        int[] digitFrequency = Maths.getDigitFrequency(n);
        int[] digits = Maths.getDigits(n);
        int unique = 0;
        for (int count : digitFrequency) {
            if (count > 0) {
                unique++;
            }
        }
        int[][] digitGroups = new int[unique][];
        int idx = 0;
        for (int curDigit = 0; curDigit < digitFrequency.length; curDigit++) {
            int count = digitFrequency[curDigit];
            if (count > 0) {
                digitGroups[idx] = new int[count];
                int groupIdx = 0;
                for (int digitIdx = 0; digitIdx < digits.length; digitIdx++) {
                    int digit = digits[digitIdx];
                    if (digit == curDigit) {
                        digitGroups[idx][groupIdx++] = digitIdx;
                    }
                }
                idx++;
            }
        }
        return digitGroups;
    }
}

File: cpp/S051.cpp
#include <iostream>
#include "primes.cpp"
#include "maths.cpp"

using namespace std;

static const int LIMIT = 1000 * 1000;
static const int TARGET = 8;

inline vector<vector<int>> get_digit_groups(int n) {
    vector<vector<int>> groups(10);
    int idx = 0;
    while (n > 0) {
        groups[n % 10].push_back(idx++);
        n /= 10;
    }
    return groups;
}

inline int replace(int n, int digit, const vector<int>& digit_group) {
    int reversed = 0, remaining = n;
    int digit_idx = 0, group_idx = 0;
    while (remaining > 0) {
        reversed *= 10;
        if (group_idx < digit_group.size() && digit_idx == digit_group[group_idx]) {
            reversed += digit;
            group_idx++;
        } else {
            reversed += remaining % 10;
        }
        digit_idx++;
        remaining /= 10;
    }
    return reverse_10(reversed);
}

inline bool is_eight_prime_fam(const vector<bool>& sieve, int n, const vector<int>& digit_group) {
    int count = 0;
    int n_digits = log_10(n) + 1;
    for (int digit = 0; digit < 10; digit++) {
        // do not replace most significant digit with 0
        if (digit_group.back() == n_digits - 1 && digit == 0) {
            continue;
        }
        count += sieve[replace(n, digit, digit_group)];
        if (digit - count >= (10 - TARGET)) {
            return false;
        }
    }
    return count == TARGET;
}

inline bool is_eight_prime_fam(vector<bool>& sieve, int n) {
    vector<vector<int>> groups = get_digit_groups(n);
    for (auto group : groups) {
        if (!group.empty() && is_eight_prime_fam(sieve, n, group)) {
            return true;
        }
    }
    return false;
}

int main() {
    vector<bool> sieve = compute_sieve(LIMIT);
    for (int i = 1; i < LIMIT; i++) {
        if (sieve[i] && is_eight_prime_fam(sieve, i)) {
            cout << i << endl;
            break;
        }
    }
    return 0;
}
